### DISCUSSION TITLE
CompleteMultipartUpload fails with InvalidPart

### DISCUSSION THREAD

**Discussion by sparkblaze** [minio/minio#16770] [Q&A]
This doesn't happen regularly, but it does happen consistently with some specific types of files that are fairly large.

> InvalidPart: One or more of the specified parts could not be found. The part may not have been uploaded, or the specified entity tag may not match the part's entity tag.

When I look back through the audit log for related events for the key, then usually there's one part that fails with a 503, but is retried and uploads successfully afterwards.

I've tried using the `aws s3api` tools to get and recreate a `CompleteMultipartUpload` request and submit it, but it fails the same with an `InvalidPart` error... is it possible to work out what that `InvalidPart` actually is if it looks like everything uploaded correctly?

>minio version RELEASE.2023-02-27T18-10-45Z

**Comment by klauspost** [comment]
We have not observed this, but that doesn't mean there isn't a problem. After all failed parts are rather rare.

The most likely is that an etag doesn't match the returned value. If you can create a clean reproducer that would speed up things quite a bit since reproducing this can be quite a goose chase without knowing more.

**Reply by sparkblaze** [reply]
Hmm, looks like I can replicate it using `mc cp` with the same files that fail - they're around 1TB.

Using `mc cp` forces the following errors after a lot longer than Rubrik could manage with the same files:

> mc.exe: <ERROR> Failed to copy `.\Clarity_20230305_0103.bak`. Put "https://minio-backups.:9000/test/Clarity_20230305_0103.bak?partNumber=275&uploadId=MDk2M2Q0ZTEtZGM4Yi00YWJmLWI0MjYtNDY4M2ZlY2YwNWY5LjE2M2NkNDM3LTMxZWMtNDNkNi1hZjkxLTc1OWM4YWYwMWMzYQ": dial tcp 192.168.202.18:9000: i/o timeout

**192.168.202.18 - minio.service**
> read tcp 192.168.202.18:9000-\u003e192.168.104.95:14246: read: connection reset by peer (*net.OpError)
> read tcp 192.168.202.18:9000-\u003e192.168.104.95:14245: read: connection reset by peer (*net.OpError)
> read tcp 192.168.202.18:9000-\u003e192.168.104.95:14244: read: connection reset by peer (*net.OpError)
> read tcp 192.168.202.18:9000-\u003e192.168.104.95:14243: read: connection reset by peer (*net.OpError)
> http://(syslog-host):8081 returned 'Post \"http://(syslog-host):8081\": lookup (syslog-host): i/o timeout', please check your endpoint configuration (*fmt.wrapError)

When Rubrik hits the same issue, it seems to throw out the `CompleteMultipartUpload` request, but when the issues are hit with `mc cp`, then that throws an `AbortMultipartUpload` which I guess makes a lot more sense.

I'm assuming that the lookup error for the syslog host is related to the errors above, but given those, where should I be looking for a resolution, at this point during the day, the cluster should have a fairly minimal load (aside from ILM policies) as we don't offload anything to it during business hours. 

I have manually provided a value for max api requests (which is well below the value that minio auto-configured for itself). Would it be worth lowering it even further if I'm using spinning disks or should I be focusing somewhere around networking?
```
#API Request Throttling
MINIO_API_REQUESTS_MAX=400
```

**Reply by Mionsz** [reply]
Hello. I am also having issues with multipart uploads - seems like sometimes some parts are not uploaded to MinIO when using Python3 based API upload - 95% of the time this is the following stack, Python:
```
boto3==1.21.7
botocore==1.24.7
```
MinIO:
```
quay.io/minio/minio:RELEASE.2022-06-11T19-55-32Z
```
MinIO after version advance 5-6 days ago:
```
quay.io/minio/minio:RELEASE.2023-04-13T03-08-07Z
```

Running inside Kubernetes environment with custom Helm Chart deployment (developed probably before one show in MinIO repo). Simple settings, no TLS nor SSH, 4 pods each with 4 node-local hdd. 
I will send a little of stack-trace in form of logs later on. This does not happen always, only sometimes and for more intense workloads, like 100Gi++ of data uploaded by many pods at the same time. What is worrying is that the failed task does not trigger error on uploader side and this ends in missing files on MinIO end. 

The files that are being send are `parquet` data format. 

Also the version advance of MinIO does not fix the issues.

**Comment by kasir-barati** [comment]
@klauspost I have a small reproducible repo for you, do not know if this is actually what @sparkblaze is experiencing but I get similar error message: https://github.com/kasir-barati/bugs/tree/aws-s3-multipart-completion-error

**Reply by kasir-barati** [reply]
BTW I have formulated my question in this stackoverflow Q&A, so if you know what is the problem I would love to learn from you: https://stackoverflow.com/questions/79440191/invalid-checksum-when-using-multipart-upload-with-full-object-checksum-in-nodejs

**Reply by klauspost** [reply]
`209188370` is not a properly encoded checksum. It must be a base64 (standard) encoded base 64 little endian value of the number.

**Reply by kasir-barati** [reply]
@klauspost Yeah, actually I moved on from that issue and now I am hitting another issue, my checksum is different from what Minio/S3 is returning. And the interesting part is that when I use the same `generateChecksum` function to generate a checksum for each chunk it works OK and what it generates is exactly the same as what Minio is generating.

But my question now is that even though I've said I wanted to have a FULL_OBJECT checksum validation my checksum of the entire file does not match the one generated by Minio. What I am missing?

I have asked the same question [here](https://repost.aws/questions/QU5H7iryj4S3q4u5-UCsmasQ/xamzcontentchecksummismatch-the-provided-x-amz-checksum-header-does-not-match-what-was-computed) too. So basically I am not sure what I am missing. Can you take a look at my [supper simple example here](https://github.com/kasir-barati/bugs/blob/348f205ad251c80ac15ef4eee27fd0e9da92b6bd/simplified.ts#L87-L89) and tell me what I am doing wrong? I can see different checksums being logged in my terminal and I wonder what I am missing about this whole topic of checksums. I do not believe that Minio/S3 calculate a checksum of checksums since I have said explcitly [here](https://github.com/kasir-barati/bugs/blob/348f205ad251c80ac15ef4eee27fd0e9da92b6bd/simplified.ts#L41) that I want a FULL_OBJECT checksum check.

**Reply by kasir-barati** [reply]
It seems like it does not matter if I say `COMPOSITE` or `FULL_OBJECT` when I am calling `CreateMultipartUploadCommand`. AWS will return the same checksum hash for my object. Look at [this commit](https://github.com/kasir-barati/bugs/commit/2aa7186afd894fe1aa7c7e62effa6d069204aed8) to understand what I mean.

You can follow the instructions [here](https://github.com/kasir-barati/bugs/tree/aws-s3-multipart-completion-error?tab=readme-ov-file#how-to-run-the-simplifiedts) to see it for yourself, try to change that value to `FULL_OBJECT` or `COMPOSITE`. It does not change the returned checksum from S3.

Any idea?

**Reply by kasir-barati** [reply]
Coming back to this thread of thoughts, so it seems that [`@aws-sdk/client-s3`](https://www.npmjs.com/package/@aws-sdk/client-s3) is waiting for AWS S3 to add `x-amz-checksum-type` header to their API as it's been discussed [here](https://github.com/aws/aws-sdk-js-v3/issues/6742#issuecomment-2547595379). But I would like to know where Minio stands about these improvements in AWS S3's improvements.

**Reply by klauspost** [reply]
@kasir-barati This is completely unrelated to MinIO. And as much as I'd like to help you fix your software, I neither have the setup or time to do that.

> But I would like to know where Minio stands about these improvements in AWS S3's improvements.

We support all of them.

You are welcome to check my code for my [minio-go implementation](https://github.com/minio/minio-go/pull/2026), which was written based on the available documentation and validated against AWS - or you can take a look at the [serverside implementation](https://github.com/minio/minio/pull/20855).

**Reply by kasir-barati** [reply]
OK, I am on it.

I realized something funny. When I upload a picture and I am using AWS Console to do so and enter a pre-calculated hash with this command: `shasum -a 256 me.jpg | cut -f1 -d\ | xxd -r -p | base64` I can upload it no problem. 

**But** even though my code is generating same checksum hash it gets a mismatch checksum error :cactus:. So I believe I am doing something fundamentally wrong. BTW for those of you interested in this you can see my code [here](https://github.com/kasir-barati/bugs/blob/ee1529360d30b15434f12a74611b3316623d2149/put.ts).

**Reply by klauspost** [reply]
SHA256 cannot be full object. So it will only match on PutObject - not multipart.

**Reply by kasir-barati** [reply]
Yeah, I also tried it just to see what AWS S3 will show me. You can only use CRC-base algorithms.

And just for future reference; my issue has been solved. Learn more here: https://github.com/minio/minio/issues/20944

**Comment by kasir-barati** [comment]
@sparkblaze based on what I am experiencing right now it does not happend as long as your checksums are correct. If you are multipart uploading use this small NodeJS script I have written [here](https://github.com/kasir-barati/bugs/blob/aws-s3-multipart-completion-error/simplified.ts) and then [you can uncomment this line](https://github.com/kasir-barati/bugs/blob/348f205ad251c80ac15ef4eee27fd0e9da92b6bd/simplified.ts#L57) and [this line](https://github.com/kasir-barati/bugs/blob/348f205ad251c80ac15ef4eee27fd0e9da92b6bd/simplified.ts#L66) to generate the checksum for each part. Though as I have asked for here [here](https://repost.aws/questions/QU5H7iryj4S3q4u5-UCsmasQ/xamzcontentchecksummismatch-the-provided-x-amz-checksum-header-does-not-match-what-was-computed) already if you try to validate the `FULL_OBJECT` checksum it will fail. IDK why as I have explained it [here](https://github.com/minio/minio/discussions/16770#discussioncomment-12208938) this should not be happneing.
